sky/serve/replica_managers.py

```diff
@@ -692,25 +736,48 @@ def _refresh_process_pool(self) -> None:
                 # initial_delay_seconds is not supported. We should add it
                 # later when we support `sky serve update`.
+                removal_reason = None
                 if info.status_property.is_scale_down_succeeded(
-                        self._initial_delay_seconds):
+                        self._get_initial_delay_seconds(info.version)):
                     # This means the cluster is deleted due to
                     # a scale down or the cluster is recovering
                     # from preemption. Delete the replica info
                     # so it won't count as a replica.
-                    serve_state.remove_replica(self._service_name, replica_id)
                     if info.status_property.preempted:
                         removal_reason = 'for preemption recovery'
                     else:
                         removal_reason = 'normally'
-                    logger.info(f'Replica {replica_id} removed from the '
-                                f'replica table {removal_reason}.')
+                # Don't keep failed record for version mismatch replicas,
+                # since user should fixed the error before update.
+                elif info.version != self.latest_version:
+                    removal_reason = 'for version outdated'
                 else:
                     logger.info(f'Termination of replica {replica_id} '
                                 'finished. Replica info is kept since some '
                                 'failure detected.')
                     serve_state.add_or_update_replica(self._service_name,
                                                       replica_id, info)
+                if removal_reason is not None:
+                    serve_state.remove_replica(self._service_name, replica_id)
+                    logger.info(f'Replica {replica_id} removed from the '
+                                f'replica table {removal_reason}.')
+
+                # Clean old version
+                replica_infos = serve_state.get_replica_infos(
+                    self._service_name)
+                current_oldest_version = min([
+                    info.version for info in replica_infos
+                ]) if replica_infos else self.oldest_version
+                if self.oldest_version < current_oldest_version:
+                    for version in range(self.oldest_version,
+                                         current_oldest_version):
+                        task_yaml = self._get_yaml(version)
+                        # Delete old version metadata.
+                        serve_state.delete_version(self._service_name, version)
+                        # Delete storage buckets of older versions.
+                        service.cleanup_storage(task_yaml)
+                    # newest version will be cleaned in serve down
+                    self.oldest_version = current_oldest_version
```

Comment:
Should we only do that for the version of the terminated replica?

Reply:
I think you would still need to check whether the terminated replica is the only remaining one of the old version.

Reply:
Yes, check the current terminated replica is the only remaining one should be enough?
Also, could we move this outside the for loop

Reply:
Moved outside the for loop.
I think checking the two are similar, because you still have to check whether it is the only remaining _one on each termination_, and how should we set the new `oldest_version`. (e.g. old_replica_version, version_with_no_replica, ..., ..., current_version)
I think let's just do it once outside the for loop.